### PR TITLE
notify_telegram plugin

### PR DESCRIPTION
**Pull request by YanikusGG**
This is initial implementation for **notify_telegram** plugin. Later I will create PR for collectd 6 also.

The **notify_telegram** plugin uses _libcurl_ and Telegram Bot API to send notifications to a configured telegram users.
It also can send config instruction with ChatID in response to all users by long polling or webhooks using _libcurl_, _libyajl_, _libmicrohttpd_.

ChangeLog: Introduce notify_telegram plugin

Reference: https://core.telegram.org/bots/api

### CODE REVIEW

File: src/notify_telegram.c [hunk 1]
```diff
@@ -0,0 +1,871 @@
...
+      return -1;
+    }
+    recipients_len++;
+  } else if (strcasecmp(key, "BotToken") == 0) {
+    sfree(bot_token);
+    bot_token = strdup(value);
+    if (bot_token == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
```
Comment (eero-t) [inline src/notify_telegram.c:732]: Code would be easier to read with a table containing the string and variable (address) pair, and a loop handling all of those.

File: src/notify_telegram.c [hunk 2]
```diff
@@ -0,0 +1,871 @@
...
+  } else if (strcasecmp(key, "WebhookPort") == 0) {
+    sfree(webhook_port);
+    webhook_port = strdup(value);
+    if (webhook_port == NULL) {
+      ERROR("notify_telegram: strdup failed.");
+      return -1;
+    }
+  } else if (strcasecmp(key, "MHDDaemonHost") == 0) {
+#if MHD_VERSION >= 0x00090000
```
Comment (eero-t) [inline src/notify_telegram.c:764]: Checks for earlier MHD version keys could be done separately from the strdup() loop.  Or the looped table could have member indicating for which MHD version give key is.

File: src/notify_telegram.c [hunk 3]
```diff
@@ -0,0 +1,871 @@
...
+
+static int notify_telegram_notification(const notification_t *n,
+                                        user_data_t __attribute__((unused)) *
+                                            user_data) {
+  char buf[MAX_BUF_SIZE] = "";
+  char *buf_ptr = buf;
+  int buf_len = sizeof(buf);
+  int status;
+
+  status = snprintf(
```
Comment (eero-t) [inline src/notify_telegram.c:800]: While it's both count and error status, I think here `count` would be more appropriate name.

File: src/notify_telegram.c [hunk 4]
```diff
@@ -0,0 +1,871 @@
...
+  }
+
+#define APPEND(bufptr, buflen, key, value)                                     \
+  if ((buflen > 0) && (strlen(value) > 0)) {                                   \
+    status = snprintf(bufptr, buflen, ",\n%s = %s", key, value);               \
+    if (status > 0) {                                                          \
+      bufptr += status;                                                        \
+      buflen -= status;                                                        \
+    }                                                                          \
+  }
```
Comment (eero-t) [inline src/notify_telegram.c:819]: Use static function instead.  Compiler can then optimize it as it sees best (inline it like you did here, or not).

File: src/notify_telegram.c [hunk 5]
```diff
@@ -0,0 +1,871 @@
...
+      buf_ptr, buf_len, "<b>Notification:</b>\nseverity = %s",
+      (n->severity == NOTIF_FAILURE)
+          ? "FAILURE"
+          : ((n->severity == NOTIF_WARNING)
+                 ? "WARNING"
+                 : ((n->severity == NOTIF_OKAY) ? "OKAY" : "UNKNOWN")));
+  if (status > 0) {
+    buf_ptr += status;
+    buf_len -= status;
+  }
```
Comment (eero-t) [inline src/notify_telegram.c:810]: You could assign severity to a separate string variable, use the append function also for this, and add `,\n` to the latter key values.

PS. I wonder though why both `\n` and `,` separators are needed, isn't one of them enough?

**Comment by eero-t** [comment]
> This is initial implementation for **notify_telegram** plugin. Later I will create PR for collectd 6 also.

New plugins should go to `collectd-6.0` branch, only existing plugins are updated in v5.x code (`main`) .

File: src/notify_telegram.c [hunk 6]
```diff
@@ -0,0 +1,871 @@
...
+#endif
+
+static int notify_telegram_init(void) {
+  curl_global_init(CURL_GLOBAL_SSL);
+
+  if (disable_getting_updates) {
+    return 0;
+  }
+
+  if (webhook_host && !httpd) {
```
[the remark this answers is not in the document]
Comment (YanikusGG) [reply src/notify_telegram.c]: thank you, I've moved the common parts of the code into separate functions

File: src/notify_telegram.c [hunk 7]
```diff
@@ -0,0 +1,871 @@
...
+}
+
+static int tg_parse_number_callback(void *ctx, const char *number,
+                                    yajl_len_t number_len) {
+  struct parse_context_t *parse_context = (struct parse_context_t *)ctx;
+
+  if (parse_context->inside_key_update_id) {
+    if (parse_context->max_update_id == NULL) {
+      parse_context->max_update_id = strndup(number, number_len);
+      if (parse_context->max_update_id == NULL) {
```
[the remark this answers is not in the document]
Comment (YanikusGG) [reply src/notify_telegram.c:158]: I tried early-exit code style and at this case it got worse (in my opinion). So I want to leave it as it is.

File: src/notify_telegram.c (hunk 1 above)
Comment (YanikusGG) [reply src/notify_telegram.c]: I don't think it's worth it, because not all keys in the config are handled the same way

**Comment by YanikusGG** [comment]
> > This is initial implementation for **notify_telegram** plugin. Later I will create PR for collectd 6 also.
> 
> New plugins should go to `collectd-6.0` branch, only existing plugins are updated in v5.x code (`main`) , but please handle review feedback first.

Thank you for your helpful issues. I've fixed most of them.

File: src/notify_telegram.c (hunk 1 above)
Comment (eero-t) [reply src/notify_telegram.c]: Most keys are, but indeed there would be few that would need separate handling outside of the proposed loop.  I'll resolve this.

File: src/notify_telegram.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (eero-t) [reply src/notify_telegram.c:158]: Thanks for testing it!

The early-exit change you did for `notify_telegram_read()` IMHO really improved / was important for it, but this function is small enough that if/else ladders are fine => resolved.

**Comment by eero-t** [comment]
Thanks, fixes look good!    Next step is review from Florian (@octo) for the v6 version of the PR.

(He hasn't been active for a while, but hopefully you don't need to wait too long...)

**Comment by eero-t** [comment]
With the v6 PR (https://github.com/collectd/collectd/pull/4311) and new plugins not being accepted to v5, closing this.

**Pull request closed without merging**